src/libopensc/pkcs15-starcos-esign_qes.c

```diff
@@ -0,0 +1,366 @@
+	sc_context_t * ctx;
+	sc_card_t * card;
+	const char * label = name_Card;
+    int r, apps_added;
+
+    if (!p15card || ! p15card->card || !p15card->card->ctx ) return SC_ERROR_INVALID_ARGUMENTS;
+
+	// convenience variables
+	card = p15card->card;
+	ctx = card->ctx;
+	
+    SC_FUNC_CALLED(ctx, SC_LOG_DEBUG_NORMAL);
+
+	if (card->type!=SC_CARD_TYPE_STARCOS_V3_4 && card->type!=SC_CARD_TYPE_STARCOS_V3_5) {
+        sc_log(p15card->card->ctx, "Unsupported card type: %d", card->type);
+        return SC_ERROR_WRONG_CARD;
+    }
+
+	apps_added = 0;
+	if ( aid == NULL ) {
+		// no aid: in this case all emulated apps are added
+		r = starcos_add_esign_app(p15card);
+		if ( r == SC_SUCCESS ) apps_added++;
+
+		r = starcos_add_qes_app(p15card);
+		if ( r == SC_SUCCESS ) apps_added++;
+	} else {
+		// aid specified: only the matching app is added
+		if (aid->len == sizeof(aid_QES) && memcmp(aid->value, aid_QES, sizeof(aid_QES)) == 0 ) {
+			r = starcos_add_qes_app(p15card);
+			if ( r == SC_SUCCESS ) {
+				label = name_QES;
+				apps_added++;
+			}
+		} else if (aid->len == sizeof(aid_ESIGN) && memcmp(aid->value, aid_ESIGN, sizeof(aid_ESIGN)) == 0 ) {
+			r = starcos_add_esign_app(p15card);
+			if ( r == SC_SUCCESS ) {
+				label = name_ESign;
+				apps_added++;
+			}
+		}
+
+		if ( apps_added == 1 ) {
+			// pkcs11 requires the file_app
+			struct sc_path  path;
+			struct sc_file *file = NULL;
+			sc_log(p15card->card->ctx, "Selecting application DF");
+			sc_path_set(&path, SC_PATH_TYPE_DF_NAME, aid->value, aid->len, 0, 0);
+			r = sc_select_file(card, &path, &file);
+			if (r != SC_SUCCESS || !file)
+				return SC_ERROR_INTERNAL;
+			sc_file_free(p15card->file_app);
+			p15card->file_app = file;		
+		}
+	}
+
+	if ( apps_added == 0 ) {
+		LOG_TEST_RET(ctx, SC_ERROR_WRONG_CARD, "Neither ESign nor QES initialized");
+	}
+
+    if (p15card->tokeninfo) {
+		sc_pkcs15_free_tokeninfo(p15card->tokeninfo);
+	}
+
+    p15card->tokeninfo = sc_pkcs15_tokeninfo_new();
+    if (!p15card->tokeninfo) {
+		LOG_TEST_RET(ctx, SC_ERROR_OUT_OF_MEMORY, "unable to create tokeninfo struct");
+	} else {
+        sc_serial_number_t serial;
+		char serial_hex[SC_MAX_SERIALNR*2+2];
+	    r = sc_card_ctl(card, SC_CARDCTL_GET_SERIALNR, &serial);
+		LOG_TEST_RET(ctx, r, "Failed to query card serial number");
+
+		sc_bin_to_hex(serial.value, serial.len , serial_hex, sizeof serial_hex, 0);
+		p15card->tokeninfo->serial_number = strdup(serial_hex);
+		p15card->tokeninfo->label = strdup(label);
+		p15card->tokeninfo->manufacturer_id = strdup(name_Vendor);
+		p15card->tokeninfo->flags = SC_PKCS15_TOKEN_READONLY;
+	}
+
+    return SC_SUCCESS;
+}
+
+int sc_pkcs15emu_starcos_esign_qes_init_ex(sc_pkcs15_card_t *p15card, struct sc_aid *aid) {
+	int r = SC_ERROR_WRONG_CARD;
+
+    if (!p15card || ! p15card->card || !p15card->card->ctx) return SC_ERROR_INVALID_ARGUMENTS;
+
+    r = starcos_esign_qes_init(p15card, aid);
```

Comment:
I suppose that StarCOS is quite flexible regarding the card profile. It looks like this emulator assumes that all 3.4/3.5 cards have the QES profile. And if they have the QES app installed, then it is assumed that the card doesn't have any other applications. Is it possible to differentiate your card profile from others? Most vendors use a specific ATR, for example...

An other problem is that you most likely don't need the generic profile parsing with `sc_pkcs15_bind_synthetic()`. If you add your card to `sc_pkcs15_is_emulation_only()`, then your profile is loaded right away even if `try_emulation_first` is not set in `opensc.conf`...

Reply:
You are right these cards should have a specific ATR, I can check for them and introduce a STARCOS subtype with this profile.

On the other hand I have tested it with a v3.4 D-Trust card which has completely different profile. In this case the emulator reports WRONG_CARD so even if try_emulation_first is set, the card would work with the PKCS15 info on the card. But if either one of the supported apps are card then it would activate the emulator (in case of PKCS15 or minidriver, the PKCS11 Framework uses a different approach by loading all apps from EF.DIR)

Reply:
I don't know what hinders the D-Trust card to be loaded, but please make the emulation profile specific for your card only.